File: vulnagent/src/llm.js
/**
 * llm.js — LLM client with two wire styles (LLM_API_STYLE):
 *
 *   anthropic  POST {baseUrl}/messages   (Anthropic Messages gateways)
 *   openai     POST {baseUrl}/chat/completions  (OpenAI-compatible gateways,
 *              e.g. opencode go for deepseek-v4-flash; Anthropic-format
 *              messages/tools are converted both ways so callers never
 *              notice)
 *
 * Output cap is per-model: deepseek-v4-flash is clamped at 4096 on some
 * gateways (and ignores thinking:disabled on hard prompts — observed
 * 2026-08-06), deepseek-v4-pro allows 8192. Model/temperature/max_tokens come
 * from .env (LLM_MODEL / LLM_TEMPERATURE / LLM_MAX_TOKENS) when set, else
 * from agent.json's model block (session.js resolves the priority).
 * Retries 429/5xx/network errors with exponential backoff.
 */
export class LLMError extends Error {
  constructor(message, { status, body } = {}) {
    super(message);
    this.status = status;
    this.body = body;
  }
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

export class LLMClient {
  constructor({ baseUrl, apiKey, model, style = "anthropic", thinking = false, maxTokens = 4096, timeoutMs = 120000, temperature = 0.2 }) {
    if (!apiKey) throw new LLMError("LLM_API_KEY is empty");
    this.baseUrl = baseUrl;
    this.apiKey = apiKey;
    this.model = model;
    this.style = style;
    this.thinking = thinking;
    this.maxTokens = maxTokens;
    this.timeoutMs = timeoutMs;
    this.temperature = Number.isFinite(Number(temperature)) ? Number(temperature) : 0.2;
    this.usage = { requests: 0, errors: 0, input_tokens: 0, output_tokens: 0 };
  }

  /**
   * One Messages API call.
   * @param {object} p {system, messages, tools, maxTokens}
   * @returns the raw response JSON (content blocks may include thinking/text/tool_use)
   */
  async messages({ system, messages, tools, maxTokens }) {
    if (this.style === "openai") {
      return this._chatCompletions({ system, messages, tools, maxTokens });
    }
    const body = {
      model: this.model,
      max_tokens: maxTokens ?? this.maxTokens,
      temperature: this.temperature,
      messages,
      ...(system ? { system } : {}),
      ...(tools?.length ? { tools } : {}),
    };
    if (!this.thinking) body.thinking = { type: "disabled" };

    const maxAttempts = 4;
    let lastErr = null;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
      try {
        const resp = await fetch(`${this.baseUrl}/messages`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            "x-api-key": this.apiKey,
            Authorization: `Bearer ${this.apiKey}`,
            "anthropic-version": "2023-06-01",
          },
          body: JSON.stringify(body),
          signal: ctrl.signal,
        });
        const text = await resp.text();
        if (resp.ok) {
          const data = JSON.parse(text);
          this._track(data);
          return data;
        }
        const retryable = resp.status === 429 || resp.status >= 500;
        lastErr = new LLMError(`LLM HTTP ${resp.status}: ${text.slice(0, 300)}`, {
          status: resp.status,
          body: text.slice(0, 1000),
        });
        if (!retryable) {
          this.usage.errors += 1;
          throw lastErr;
        }
      } catch (err) {
        if (err instanceof LLMError) throw err;
        lastErr = new LLMError(`LLM request failed: ${err.message}`);
      } finally {
        clearTimeout(timer);
      }
      if (attempt < maxAttempts) await sleep(1000 * 2 ** attempt);
    }
    this.usage.errors += 1;
    throw lastErr ?? new LLMError("LLM request failed");
  }

  // ------------------------------------------------------------------
  // OpenAI chat-completions style (opencode go and other OAI gateways).
  // Anthropic-format messages/tools in, Anthropic-shaped response out, so
  // session.js never notices the difference.
  // ------------------------------------------------------------------

  static _toOpenAI(messages) {
    const out = [];
    for (const m of messages) {
      const content = m.content;
      if (typeof content === "string") {
        out.push({ role: m.role, content });
        continue;
      }
      const blocks = Array.isArray(content) ? content : [];
      if (m.role === "assistant") {
        const text = blocks.filter((b) => b.type === "text" && b.text)
          .map((b) => b.text).join("\n");
        const toolCalls = blocks.filter((b) => b.type === "tool_use")
          .map((b) => ({
            id: b.id,
            type: "function",
            function: { name: b.name, arguments: JSON.stringify(b.input ?? {}) },
          }));
        out.push({
          role: "assistant",
          content: text || null,
          ...(toolCalls.length ? { tool_calls: toolCalls } : {}),
        });
        continue;
      }
      // user role: tool_result blocks become role:"tool" messages
      const toolResults = blocks.filter((b) => b.type === "tool_result");
      const texts = blocks.filter((b) => b.type === "text" && b.text);
      for (const tr of toolResults) {
        const body = Array.isArray(tr.content)
          ? tr.content.filter((b) => b.type === "text").map((b) => b.text).join("\n")
          : String(tr.content ?? "");
        out.push({ role: "tool", tool_call_id: tr.tool_use_id, content: body });
      }
      if (texts.length || !toolResults.length) {
        out.push({ role: "user", content: texts.map((b) => b.text).join("\n") });
      }
    }
    return out;
  }

  static _fromOpenAI(data) {
    const choice = data.choices?.[0] ?? {};
    const msg = choice.message ?? {};
    const content = [];
    if (msg.content) content.push({ type: "text", text: msg.content });
    for (const tc of msg.tool_calls ?? []) {
      let input = {};
      try { input = JSON.parse(tc.function?.arguments || "{}"); } catch { /* keep {} */ }
      content.push({
        type: "tool_use",
        id: tc.id,
        name: tc.function?.name,
        input,
      });
    }
    const finishMap = { stop: "end_turn", tool_calls: "tool_use", length: "max_tokens" };
    return {
      content,
      stop_reason: finishMap[choice.finish_reason] ?? choice.finish_reason ?? "end_turn",
      usage: {
        input_tokens: data.usage?.prompt_tokens ?? 0,
        output_tokens: data.usage?.completion_tokens ?? 0,
      },
    };
  }

  async _chatCompletions({ system, messages, tools, maxTokens }) {
    const body = {
      model: this.model,
      max_tokens: maxTokens ?? this.maxTokens,
      temperature: this.temperature,
      messages: [
        ...(system ? [{ role: "system", content: system }] : []),
        ...LLMClient._toOpenAI(messages),
      ],
      ...(tools?.length ? {
        tools: tools.map((t) => ({
          type: "function",
          function: { name: t.name, description: t.description, parameters: t.input_schema },
        })),
      } : {}),
    };
    const maxAttempts = 4;
    let lastErr = null;
    for (let attempt = 1; attempt <= maxAttempts; attempt++) {
      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), this.timeoutMs);
      try {
        const resp = await fetch(`${this.baseUrl}/chat/completions`, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${this.apiKey}`,
          },
          body: JSON.stringify(body),
          signal: ctrl.signal,
        });
        const text = await resp.text();
        if (resp.ok) {
          const data = LLMClient._fromOpenAI(JSON.parse(text));
          this._track(data);
          return data;
        }
        const retryable = resp.status === 429 || resp.status >= 500;
        lastErr = new LLMError(`LLM HTTP ${resp.status}: ${text.slice(0, 300)}`, {
          status: resp.status,
          body: text.slice(0, 1000),
        });
        if (!retryable) {
          this.usage.errors += 1;
          throw lastErr;
        }
      } catch (err) {
        if (err instanceof LLMError) throw err;
        lastErr = new LLMError(`LLM request failed: ${err.message}`);
      } finally {
        clearTimeout(timer);
      }
      if (attempt < maxAttempts) await sleep(1000 * 2 ** attempt);
    }
    this.usage.errors += 1;
    throw lastErr ?? new LLMError("LLM request failed");
  }

  _track(data) {
    const u = data.usage ?? {};
    this.usage.requests += 1;
    this.usage.input_tokens += Number(u.input_tokens ?? 0);
    this.usage.output_tokens += Number(u.output_tokens ?? 0);
  }
}

File: fwgraph/pipeline/llm.py
"""Minimal JSON chat client. Honors LLM_API_STYLE=openai|anthropic."""

from __future__ import annotations

import json
import os
import re

import httpx
from tenacity import retry, stop_after_attempt, wait_exponential_jitter

_JSON_FENCE = re.compile(r"```(?:json)?\s*(.*?)\s*```", re.DOTALL)


def enabled() -> bool:
    return bool(os.getenv("LLM_API_KEY", "").strip())


def _style() -> str:
    return os.getenv("LLM_API_STYLE", "openai").strip().lower()


def _endpoint() -> str:
    base = os.getenv("LLM_BASE_URL", "https://api.openai.com/v1").rstrip("/")
    if _style() == "anthropic":
        if base.endswith("/messages"):
            return base
        return base + "/messages"
    if base.endswith("/chat/completions"):
        return base
    return base + "/chat/completions"


def _max_tokens(default: int) -> int:
    raw = os.getenv("LLM_MAX_TOKENS", "").strip()
    if not raw:
        return default
    try:
        return max(256, int(raw))
    except ValueError:
        return default


def parse_json(text: str):
    raw = (text or "").strip()
    if not raw:
        raise ValueError("empty LLM content")
    fence = _JSON_FENCE.search(raw)
    if fence:
        raw = fence.group(1).strip()
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        pass
    start_obj, start_arr = raw.find("{"), raw.find("[")
    starts = [i for i in (start_obj, start_arr) if i >= 0]
    if not starts:
        raise ValueError("LLM content is not JSON")
    start = min(starts)
    try:
        return json.loads(raw[start:])
    except json.JSONDecodeError as exc:
        raise ValueError(f"LLM JSON parse failed: {exc}") from exc


def content_from_response(body: dict) -> str:
    """Extract assistant text from OpenAI or Anthropic response JSON."""
    if not isinstance(body, dict):
        return ""
    choices = body.get("choices") or []
    if choices:
        msg = (choices[0].get("message") or {})
        return str(msg.get("content") or msg.get("reasoning_content") or "")
    texts = []
    for block in body.get("content") or []:
        if not isinstance(block, dict):
            continue
        if block.get("type") in ("text", "output_text") and block.get("text"):
            texts.append(str(block["text"]))
    return "\n".join(texts)


@retry(stop=stop_after_attempt(3), wait=wait_exponential_jitter(1, 8),
       reraise=True)
def chat_json(system: str, user: str, *, timeout: float = 180.0,
              max_tokens: int | None = None):
    """Return parsed JSON (object or array) from a single chat turn."""
    key = os.getenv("LLM_API_KEY", "").strip()
    if not key:
        raise RuntimeError("LLM_API_KEY is not set")
    model = os.getenv("LLM_MODEL", "deepseek-v4-flash")
    tokens = max_tokens if max_tokens is not None else _max_tokens(16384)
    headers = {"Authorization": f"Bearer {key}",
               "Content-Type": "application/json"}
    if _style() == "anthropic":
        payload = {
            "model": model,
            "max_tokens": tokens,
            "temperature": 0.1,
            "system": system,
            "messages": [{"role": "user", "content": user}],
        }
        if os.getenv("LLM_THINKING", "0") != "1":
            payload["thinking"] = {"type": "disabled"}
        headers["x-api-key"] = key
        headers["anthropic-version"] = "2023-06-01"
    else:
        payload = {
            "model": model,
            "temperature": 0.1,
            "max_tokens": tokens,
            "messages": [
                {"role": "system", "content": system},
                {"role": "user", "content": user},
            ],
        }
    with httpx.Client(timeout=timeout) as client:
        resp = client.post(_endpoint(), json=payload, headers=headers)
        resp.raise_for_status()
        body = resp.json()
    return parse_json(content_from_response(body))
